Feature/bz60589-1 PR for review (logkit to slf4j / log4j2)

woonsan opened the pull request.

woonsan committed:
60589: adding log4j2 related jar dependencies.

woonsan committed:
60589: Adding logging lib into compile classpath and dist jar packaging.

woonsan committed:
60589: Removing runtime scope dependencies in eclipse.classpath.

woonsan committed:
Experiment: using log4j2 instead of excalibur at runtime.

woonsan committed:
Merge branch 'trunk' into feature/bz60589-1

woonsan committed:
Removing slf4j_logkit and dep on logkit.

woonsan committed:
Removing avalon and excalibur.

woonsan commented:
This is only for review on direction at the moment. See the comment for detail:
- https://bz.apache.org/bugzilla/show_bug.cgi?id=60589#c4

woonsan committed:
60589: removing unused logging dependencies; adding new logging dependencies; fixing module version checking regex in a test case; checking log4j2.xml from the current working directory first before checking it from bin/ folder.

woonsan committed:
Fixing unit test due to logkit target usages. Replacing the test to use slf4j logger wrapper to keep the log records when testing counts.

woonsan committed:
formatting eclipse.classpath to use 4 spaces.

woonsan committed:
Merging trunk

woonsan committed:
Enabling LoggerPanel with log4j2-backed event listener mechanism.

woonsan committed:
Removing unused logkit classes.

woonsan committed:
Merge branch 'trunk' into feature/bz60589-1

woonsan committed:
Supporting log level command line argument; removing commons-logging sys prop; javadocs.

woonsan committed:
Merge branch 'trunk' into feature/bz60589-1

woonsan committed:
Removing separate jmeter category; root category for jmeter and all, which is more backward-compatible.

woonsan committed:
Supporting most logging command line args as backward compatibility.

woonsan committed:
Adding -i option to be able to specify logging config file; fixing error log event handler in UI; some javadoc/comments.

woonsan committed:
Merge branch 'trunk' into feature/bz60589-1

woonsan committed:
Moving logging properties to log4j2 xml files.

woonsan commented:
Now, I think this PR is ready for review (bugzilla: https://bz.apache.org/bugzilla/show_bug.cgi?id=60589).
Please take a review.

woonsan committed:
Merge branch 'trunk' into feature/bz60589-1

woonsan committed:
Fixing checkstyle in the logkit forked class.

pmouawad commented:
Hello ,
Thanks for this big piece of work.
I made a first review, it looks good to me.
There's one thing I don't clearly understand, what's the use of replaceDateFormatInFileName in NewDriver ?

Thanks

woonsan commented:
@pmouawad , thanks a lot for the quick review!
Regarding NewDriver#replaceDateFormatInFileName(), here's the reason why I added that:
- As LoggingManager#makeWriter() was gone (https://github.com/apache/jmeter/blob/trunk/src/jorphan/org/apache/jorphan/logging/LoggingManager.java#L174), I needed to have the same feature somewhere else to keep the backward compatibility for the log file name option with date format (http://jmeter.apache.org/usermanual/get-started.html#options)
- However, the date format in the log file name option has been broken for long time in my understanding. For example, if I pass "jmeter_'yyyyMMddHHmmss'.log" option value, it simply fails because SimpleDateFormat complains about unrecognized format character, 'j', for instance.
- Therefore, I rewrote the LoggingManager#makeWriter() method in NewDriver#replaceDateFormatInFileName() in order to extract string inside single quote pair and format / append the value using SimpleDateFormat, not the whole (which leads formatting failure always, resulting log file name option used literally, not evaluated with current date).

woonsan committed:
Merge branch 'trunk' into feature/bz60589-1

woonsan committed:
Moving logging configuration example in user.properties to log4j.xml as well.

woonsan committed:
Moving logging configuration examples for reporting in user.properties to log4j2.xml.

woonsan committed:
use tab instead of 4 sp in eclipse.classpath to merge from trunk

woonsan committed:
Merging trunk

woonsan committed:
backward compatibility - Setting log level for test element in help menus.

woonsan committed:
Leave info log on log level change through help menus.

woonsan committed:
Fixing HttpMirrorServer to use log4j env.

woonsan committed:
BC: supporting log level setting through command line args in HttpMirrorServer

woonsan committed:
Merge branch 'trunk' into feature/bz60589-1

woonsan committed:
Fix error when (accidentally) using GuiLogAppender without initializing application properties.

woonsan commented:
I've fixed more:
- Keep the backward compatibility in the Help / En(Dis)able DEBUG logging menu items.
- Fixed errors in HttpMirrorServer (executed via mirror-server.sh or mirror-server.bat). It now uses log4j. Also, it provides log level setting through command line arguments (e.g, `-LDEBUG' or `-Ljmeter=DEBUG -Ljorphan=DEBUG'). This command line arguments for log level settings are provided to replace the old log level setting functionality through .properties file loading). It supports `-P8888' to set port number command line arg as well as the backward compatible the first port number arg (e.g, ./mirror-server.sh 8888).
- Fixed a corner case when GuiLogAppender is used in log4j2.xml while JMeter application properties were not initialized properly.

woonsan committed:
Merge branch 'trunk' into feature/bz60589-1

FSchumacher commented on src/core/org/apache/jmeter/NewDriver.java:
```diff
@@ -276,4 +276,106 @@ private static String exceptionsToString(List<Exception> exceptionsInInit) {
+    /*
+     * Find command line argument option value by the id and name.
+     */
+    private static String getCommandLineArgument(String [] args, int id, String name) {
```
Why is `id` an `int`, when it is used as a `char` below, only?

woonsan replied:
I simply used the same pattern as the Avalon CLOptionDescriptor (using the id as integer for some reason) usages in JMeter.java, hoping for better improvement (merging command line args handling in both location somehow) in the future. At the moment, it's not easy because the class loader contexts are different between NewDriver's and JMeter's.

FSchumacher commented on src/core/org/apache/jmeter/gui/logging/GuiLogEventBus.java:
```diff
@@ -0,0 +1,79 @@
+    /**
+     * Registered GUI log event listeners array.
+     */
+    private GuiLogEventListener [] listeners;
```
The space between `GuiLogEventListener` and `[]` looks strange. Why not a `List` or `Collection`?

woonsan replied:
You're right. Thanks for the review. And thanks for the fix, @pmouawad

FSchumacher commented on src/core/org/apache/jmeter/gui/logging/GuiLogEventBus.java:
```diff
@@ -0,0 +1,79 @@
+     * @param logEvent log event object
+     */
+    public void postEvent(LogEventObject logEventObject) {
+        if (listeners != null) {
```
If `listeners` would have been a `List` (and always initialized), we could drop this condition.

FSchumacher commented on src/protocol/http/org/apache/jmeter/protocol/http/control/HttpMirrorServer.java:
```diff
@@ -159,14 +192,84 @@ public Exception getException(){
         serv.start();
     }
+
+    private static Logger getLogger() {
```
Is `HttpMirrorServer` used from one thread, only?

woonsan replied:
It's starting a server accepting a socket, on which it handles in multi-threaded worker env. Anyway, the logger has been a static member shared by all. Shouldn't be a problem.

pmouawad commented:
I'm not sure commons-logging can be dropped as it's a compile dependency of HttpClient 4.5.3:
https://repo1.maven.org/maven2/org/apache/httpcomponents/httpclient/4.5.3/httpclient-4.5.3.pom

FSchumacher commented on src/core/org/apache/jmeter/NewDriver.java:
```diff
@@ -276,4 +276,106 @@ private static String exceptionsToString(List<Exception> exceptionsInInit) {
+    }
+
+    /*
+     * If the fileName contains at least one set of paired single-quotes, reformat using DateFormat
```
I haven't found the documentation, where the format is specified for the user. Would it be simpler to use a standard time/date formatter from the jdk instead of implementing a simple parser ourselves?

woonsan replied:
@FSchumacher see the example date formatted file name example in http://jmeter.apache.org/usermanual/get-started.html#options (section 1.4.8).

pmouawad commented:
Forget about my comment on jcl

vlsi commented on src/core/org/apache/jmeter/NewDriver.java:
```diff
@@ -276,4 +276,106 @@ private static String exceptionsToString(List<Exception> exceptionsInInit) {
+    /*
+     * If the fileName contains at least one set of paired single-quotes, reformat using DateFormat
+     */
+    private static String replaceDateFormatInFileName(String fileName) {
```
`SimpleDateFormat` supports escaping via `'` (single quotes), so the whole thing could be rewritten to `return new SimpleDateFormat("'" + fileName + "'").format(new Date())`

woonsan replied:
@vlsi That's different from what we have supported now as explains in http://jmeter.apache.org/usermanual/get-started.html#options. It has an example without single quote escapes, which makes failure.

vlsi commented on src/core/org/apache/jmeter/gui/GuiPackage.java:
```diff
@@ -71,10 +72,6 @@
-                    DefaultTreeNodeNamingPolicy.class.getName());
-
     /** Singleton instance. */
     private static GuiPackage guiPack;
```
It should be volatile

woonsan replied:
I'm not sure because it's initialized once by GuiPackage.initInstance() called by JMeter#startGui() during the GUI initialization. There seems no concurrency.

ham1 commented on src/core/org/apache/jmeter/gui/logging/LogEventObject.java:
```diff
@@ -0,0 +1,75 @@
+        if (level != null) {
+            return level.isMoreSpecificThan(Level.ERROR);
+        }
+        return false;
```
These could be simplified to `return level != null && level.isMoreSpecificThan(Level.XXX);`

woonsan replied:
Right.
